Fix bug where __getstate__ of DDP looks for self._replicated_tensor_module

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #76349

When we are not using ReplicatedTensor in DDP and try to save a DDP
module it will error out since it tries to delete the _replicated_tensor_module
attribute.

Fixing this by checking if this mode is enabled before triggering the delete.

Differential Revision: [D35875167](https://our.internmc.facebook.com/intern/diff/D35875167/)

pritamdamania87 committed:
Fix bug where __getstate__ of DDP looks for self._replicated_tensor_module

When we are not using ReplicatedTensor in DDP and try to save a DDP
module it will error out since it tries to delete the _replicated_tensor_module
attribute.

Fixing this by checking if this mode is enabled before triggering the delete.

Differential Revision: [D35875167](https://our.internmc.facebook.com/intern/diff/D35875167/)

[ghstack-poisoned]

mrshenli commented on test/distributed/_shard/test_replicated_tensor.py:
```diff
@@ -285,3 +286,14 @@ def test_with_ddp(self):
+        with _ddp_replicated_tensor(False):
+            ddp = DDP(model)
+            buffer = io.BytesIO()
+            torch.save(ddp, buffer)
```
do we want to test load as well?

pritamdamania87 replied:
Added

mrshenli reviewed (approved):
Stamp to unblock

pritamdamania87 committed:
Update on "Fix bug where __getstate__ of DDP looks for self._replicated_tensor_module"


When we are not using ReplicatedTensor in DDP and try to save a DDP
module it will error out since it tries to delete the _replicated_tensor_module
attribute.

Fixing this by checking if this mode is enabled before triggering the delete.

Differential Revision: [D35875167](https://our.internmc.facebook.com/intern/diff/D35875167/)

[ghstack-poisoned]

pritamdamania87 commented:
@pytorchbot merge this please

pritamdamania87 commented:
@pytorchbot merge this please

rohan-varma commented on test/distributed/_shard/test_replicated_tensor.py:
```diff
@@ -285,3 +286,23 @@ def test_with_ddp(self):
                 self.assertIsInstance(out, ReplicatedTensor)
+
+        # Test save and load.
+        with _ddp_replicated_tensor(False):
```
Surprised we didn't have such a test in DDP, should we move this test there?
